Fix Bugs around Particle System in  Release/2.10.0

WUCJ638 opened the pull request:
# 2025.12.09 Fix Bugs around Particle System in Release/2.10.0

### Summary

This PR fixes several particle system bugs in **Axmol 2.10.0**, specifically related to:

- **SpawnFadeIn / SpawnScaleIn**:  
  - Memory not reallocated after `setTotalParticles()`, causing null pointer crashes.  
  - New values ignored after initial allocation.  
  - Particles become invisible if `_spawnFadeIn` or `_spawnScaleIn` is set to `0.0f`.

- **Loop Animation crash**:  
  - After `setTotalParticles()`, `_particleCount` wasn’t set to 0, leading to invalid access in `_animations.at(_particleData.animIndex[i])`.

### Changes

- Reallocate attribute memory (fade/scale/HSV) in `ParticleSystemQuad::setTotalParticles()` if enabled.
- Allow updating spawn fade/scale values even after allocation.
- Skip fade/scale logic when value is `0.0f` to avoid invisible particles.
- Reset `_particleCount = 0` in `setTotalParticles()` and `resetSystem()` to prevent stale animation indices.

### Testing

- Verified on **Windows 10** with Visual Studio 2026 (MSVC 19.50).
- ParticleTest from 42 to 55 pass without crash; emission shape unaffected, by simply adding

    ```cpp
        this->runAction(Sequence::create(
            DelayTime::create(3.0f),
            CallFunc::create([this](){
                _emitter->setTotalParticles(2000);
                }),
            nullptr
        ));
    ```

    at the end of `onEnter()`

- HSV & advanced animations not fully tested (because I haven't used them in my projects).

> Note: 
> 1. Huge comments (`//// BUG from WUCJ638... //`) are removed to reduce space.
> 2. Since I'm a student who is occupied with heavy academic tasks and preparing for final exam, **delayed response may be unavoidable**. Thanks for your understanding!

WUCJ638 committed:
Fix memory allocation, fade-in setter, and fade-in updating logic in ParticleSystem

Refactor memory allocation checks of particle system's attributes like fade-in, scale in, and fix bugs related to setting and updating opacity and scale fade-in. Let current particle count 0 (because all existed particles are removed)  during system reset.

WUCJ638 committed:
Enhance particle memory management on ScaleIn, FadeIn, HSV, Animation

All implementation is inside setTotalParticle(). Fix the missing allocation of scale in, opacity fade in, HSV, and animation. Meanwhile, added assertions and for particle limits

j-jorge commented:
(I'm not a maintainer)
> Note: Huge comments (//// BUG from WUCJ638... //) can be removed, since they occupy a lot of space.

I believe you should remove them indeed. Explanations for the bug and the fix are welcome in a commit message, comments about the current code's behavior are great in the code. And no need for so may slashes :)

WUCJ638 commented:
> I believe you should remove them indeed. Explanations for the bug and the fix are welcome in a commit message, comments about the current code's behavior are great in the code. And no need for so may slashes :)

Got it, I'll erase them away later on because now it's time for me to go to bed. 🛏

WUCJ638 committed:
Simplified comments inside this file

For all comments written by me, I use neater comments in place of huge comments (`/// BUG from WUCJ638: ... //`).

WUCJ638 committed:
Simplified comments inside this file

For all comments written by me, I use neater comments in place of huge comments (`/// BUG from WUCJ638: ... //`).

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -860,8 +862,8 @@ void ParticleSystem::addParticles(int count, int animationIndex, int animationCe
 
-    if (animationCellIndex != -1 || animationIndex != -1)
-        allocAnimationMem();
+     if (animationCellIndex != -1 || animationIndex != -1)
```
```suggestion
    if (animationCellIndex != -1 || animationIndex != -1)
```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -860,8 +862,8 @@ void ParticleSystem::addParticles(int count, int animationIndex, int animationCe
-    if (animationCellIndex != -1 || animationIndex != -1)
-        allocAnimationMem();
+     if (animationCellIndex != -1 || animationIndex != -1)
+         allocAnimationMem();
```
```suggestion
        allocAnimationMem();
```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -2101,31 +2107,34 @@ void ParticleSystem::useHSV(bool hsv)
 void ParticleSystem::setSpawnFadeIn(float time)
 {
-    if (time != 0.0F && !allocOpacityFadeInMem())
+// Modification is allowed when:
```
```suggestion
    // Modification is allowed when:
```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -2101,31 +2107,34 @@ void ParticleSystem::useHSV(bool hsv)
 {
-    if (time != 0.0F && !allocOpacityFadeInMem())
+// Modification is allowed when:
+//  - memory hasn't been allocated, time != 0.0F and allocation has done successfully; OR
```
```suggestion
    //  - memory hasn't been allocated, time != 0.0F and allocation has done successfully; OR
```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -2101,31 +2107,34 @@ void ParticleSystem::useHSV(bool hsv)
-    if (time != 0.0F && !allocOpacityFadeInMem())
+// Modification is allowed when:
+//  - memory hasn't been allocated, time != 0.0F and allocation has done successfully; OR
+//  - memory has allocated.
```
```suggestion
    //  - memory has allocated.
```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -118,7 +118,7 @@ bool ParticleData::init(int count)
     deltaColorB = (float*)malloc(count * sizeof(float));
     deltaColorA = (float*)malloc(count * sizeof(float));
-
+    
```
```suggestion

```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -138,6 +138,7 @@ bool ParticleData::init(int count)
     modeB.deltaRadius      = (float*)malloc(count * sizeof(float));
     modeB.radius           = (float*)malloc(count * sizeof(float));
 
+
```
```suggestion
```

halx99 commented on core/2d/ParticleSystemQuad.cpp:
```diff
@@ -694,10 +701,12 @@ void ParticleSystemQuad::setTotalParticles(int tp)
         if (!_particleData.init(tp))
         {
             AXLOGW("Particle system: not enough memory");
+            _particleData.release();    // WUCJ638
```
```suggestion
            _particleData.release();
```

halx99 commented on core/2d/ParticleSystemQuad.cpp:
```diff
@@ -686,6 +688,11 @@ void ParticleSystemQuad::draw(Renderer* renderer, const Mat4& transform, uint32_
+        "Adding more than 10000 particles will crash the renderer, the mesh generated has an index format of "
+        "U_SHORT (uint16_t).");
+
+    tp = std::min(tp, 10000);
```
```suggestion
    tp = (std::min)(tp, 10000);
```

halx99 committed:
Apply suggestions from code review

halx99 committed:
Update core/2d/ParticleSystem.cpp

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -2101,31 +2106,34 @@ void ParticleSystem::useHSV(bool hsv)
+    // Modification is allowed when:
+    //  - memory hasn't been allocated, time != 0.0F and allocation has done successfully; OR
+    //  - memory has allocated.
+    if(!_isOpacityFadeInAllocated && (time == 0.0F || !allocOpacityFadeInMem()))
```
```suggestion
    if (!_isOpacityFadeInAllocated && (time == 0.0F || !allocOpacityFadeInMem()))
```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -2101,31 +2106,34 @@ void ParticleSystem::useHSV(bool hsv)
 void ParticleSystem::setSpawnFadeInVar(float time)
 {
-    if (time != 0.0F && !allocOpacityFadeInMem())
+    if(!_isOpacityFadeInAllocated && (time == 0.0F || !allocOpacityFadeInMem()))
```
```suggestion
    if (!_isOpacityFadeInAllocated && (time == 0.0F || !allocOpacityFadeInMem()))
```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -2101,31 +2106,34 @@ void ParticleSystem::useHSV(bool hsv)
 void ParticleSystem::setSpawnScaleIn(float time)
 {
-    if (time != 0.0F && !allocScaleInMem())
+    if(!_isScaleInAllocated && (time == 0.0F || !allocScaleInMem()))
```
```suggestion
    if (!_isScaleInAllocated && (time == 0.0F || !allocScaleInMem()))
```

halx99 commented on core/2d/ParticleSystem.cpp:
```diff
@@ -2101,31 +2106,34 @@ void ParticleSystem::useHSV(bool hsv)
 void ParticleSystem::setSpawnScaleInVar(float time)
 {
-    if (time != 0.0F && !allocScaleInMem())
+    if(!_isScaleInAllocated && (time == 0.0F || !allocScaleInMem()))
```
```suggestion
    if (!_isScaleInAllocated && (time == 0.0F || !allocScaleInMem()))
```

halx99 committed:
Apply suggestions from code review
